main.go

```diff
@@ -1700,12 +1701,26 @@ func (tf *transformer) removeUnnecessaryImports(file *ast.File) {
 			return true
 		}
 
-		if pkg := uses.Pkg(); pkg != nil {
-			usedImports[pkg.Path()] = true
+		pkg := uses.Pkg()
+		if pkg == nil {
+			return true
 		}
 
+		var pkgName string
+		switch parent := cursor.Parent().(type) {
+		case *ast.SelectorExpr:
+			pkgNameIdent, ok := parent.X.(*ast.Ident)
+			if !ok {
+				return true // TODO(pagran): review this
+			}
+			pkgName = pkgNameIdent.Name
+		default:
+			pkgName = "."
+		}
```

Comment:
can you document the logic here? I'm struggling to follow it. I can sort of see what you're trying to achieve, but I can't understand the code.

Also, what is the TODO for?

Reply:
in last commit detection logic rewrited

realization through `SelectorExpr` is too hacky, i made it through `types`

Reply:
Your fix for non-dot-imports looks OK; it should handle both regular imports and renamed imports.

I'm not really following how you handle dot-imported declarations.  why isn't that matching names from non-dot-imports as well? if the scope lookup somehow solves that, please document it

Reply:
Add a comment to code. Idea from here - https://github.com/golang/go/blob/master/src/go/types/object.go#L515

Reply:
Sure, but doesn't this still match names whose package was imported _without_ a dot import?

that is, for `import "foo/bar"; var _ = bar.Baz`, when you're walking the syntax tree, the identifier `Baz` will still match this conditional. The object has a packge, which is not our current package (because it's `foo/bar`), and looking up the name in that package gives us the same `Bar`.

the code you quoted from go/types doesn't filter by dot imports, I'm pretty sure.

Reply:
Hmm, it's not intended behavior, but it doesn't affect any logic (just useless actions). For a more formal way, i should check parent node and if it's `SelectExpr` then it's a normal import, otherwise - dot-imports

Reply:
I don't think it's harmless. You're effectively treating a dot import as used in a scenario where the only "use" is like `pkg.Foo`, not `Foo`. I think you just happen to not cover that scenario in your added test case, because in your added test case _all_ three forms of the import become unused.

Reply:
In other words, I think we should have more tests, like:

1) one file that imports a regular and a renamed package, and the renamed becomes unused
2) one file that imports a regular and a dot-imported package, and the dot import becomes unused
3) one file that imports a regular and a dot-imported package, and the regular becomes unused (to double check with number 2, since dot imports are nasty)

Reply:
You're right. Second case crashes if I don't check `SelectorExpr`. Good point.

Reply:
FYI telling if the parent node is a selector expression is still not quite enough.

```
import . "foo/bar" // contains name Baz

var _ = Baz // you correctly detect this as a use of the dot-imported package
var _ = Baz.SomeField // you miss this one
```

I _think_ it might be enough to check that the parent is a SelectorExpr, and that the identifier is its _right_ side, i.e. the `Sel` field. Because for a normally imported package, we can only refer to `Baz` as `foo.Baz`.

Reply:
It's also possible that we could somehow write this detection logic with `go/types`, but I'm honestly not sure how to do it, so I think a few lines of AST detection is OK.

Reply:
But if you want to try go/types for the sake of not having hacks with go/ast, perhaps we can check if the object belongs to a _different_ package, but its scope is our _current_ package. I think that's how go/types would treat dot-imported names, because they come from a different package (the imported one) but they are added to the current package's scope.

Reply:
I think I had that almost right, with one caveat. The dot-imported names don't get added to the _current package scope_, but to the _current file scope_.

https://cs.opensource.google/go/go/+/refs/tags/go1.19.5:src/go/types/resolver.go;l=329

That makes sense, because imports affect one file, not all files in the package.

Reply:
Last thought. I think you can do this via https://pkg.go.dev/go/types#Scope.LookupParent.

Whenever you see an object, and it's not a package, then see if LookupParent on the current scope finds the _same object_. If it does, then the object was inserted into your file scope via a dot-import.

Reply:
I'm favoring ast. For go/types version it requires fill `tf.info.Scopes`. I don't think it's necessary